source/docs/software/basic-programming/functions-as-data.rst

```diff
@@ -0,0 +1,71 @@
+
+Typically, we have a need to treat functions as data when the function needs to be called from somewhere other than where it was defined in the code.
+
+For example, the Command-based framework <TODO: link> is built on ``Command`` objects that refer to methods defined on various ``Subsystem`` classes.  Many of the included ``Command`` types (such as ``InstantCommand`` and ``RunCommand``) work with *any* function - not just functions associated with a single ``Subsystem``.  To support building commands generically, we need to support passing functions from a ``Subsystem`` (which interacts with the hardware) to a ``Command`` (which interacts with the scheduler).
+
+Treating Functions as Data in Java
+----------------------------------
+
+Java represents functions-as-data as instances of "functional interfaces" <TODO: link to docs>.  A functional interface is a special kind of class that has only a single method - since Java was originally designed strictly for object-oriented programming, it has no way of representing a single function detached from a class - instead, it defines a particular group of classes that *only* represent single functions.
+
+This might sound complicated, but in the context of WPILib we don't really need to worry much about using the functional interfaces themselves - the code that does that is internal to WPILib.  Instead, all we need to know is how to pass a function that we've written to a method that takes a functional interface as a parameter.  For a simple example, consider the signature of ``Commands.runOnce`` (which creates an ``InstantCommand`` that, when scheduled, runs the given function once and then terminates):
+
+.. note:: The ``requirements`` parameter is explained in the Command-based documentation, and will not be discussed here <TODO: link>.
+.. code-block:: java
+
+   public static CommandBase runOnce(Runnable action, Subsystem... requirements)
+
+``runOnce`` expects us to give it a ``Runnable`` parameter (named ``action``).  A ``Runnable`` is the Java term for a function that takes no parameters and returns no value.  When we call ``runOnce``, we need to give it a function with no parameters and no return value.  There are two ways to do this: we can refer to some existing function using a "method reference", or we can define the function we want inline using a "lambda expression".
+
+Method References
+^^^^^^^^^^^^^^^^^
+
+A method reference lets us pass an already-existing function as our ``Runnable``:
+
+.. code-block:: java
+
+   // Create an InstantCommand that runs the `resetEncoders` method of the `drivetrain` object
+   Command disableCommand = runOnce(drivetrain::resetEncoders, drivetrain);
+
+The expression ``drivetrain::resetEncoders`` is a reference to the ``resetEncoders`` method of the ``drivetrain`` object.  It is not a method *call* - this line of code does not *itself* reset the encoders of the drivetrain.  Instead, it returns a ``Command`` that will do so *when it is scheduled.*
```

Comment:
This paragraph is absolutely critical. Students are so accustomed to invoking methods, despite the different syntax, they don't appreciate that the resetEncoders method is not being invoked at this time.

Reply:
This is a problem that extends outside of computer science; people in general tend to conflate `f` and `f(x)` all the time.  Doing this can be a pretty terrible category error and lead to all sorts of confusion.